### DISCUSSION TITLE
How to configure the serial bridge report format?

### DISCUSSION THREAD

**Discussion by kobuki** [arendst/Tasmota#22363] [Q&A] [answered]
I'm struggling with another topic regarding the ESP32 serial interface. I configured the serial bridge per the following:

```
SetOption69 1
SBaudRate 115200
SSerialConfig 8N1
SSerialSend2 1111
```

That is: invert RX input, and then some standard parameters. I receive ASCII data from our smart meter properly, something like this:

```
14:49:36.316 MQT: tele/ma105-meter/RESULT = {"SSerialReceived":"/SAG5SAG-METER\r\n\r\n0-0:1.0.0(redacted)\r\n0-0:42.0.0(redacted)\r\n0-0:96.1.0(redacted)\r\n0-0:96.14.0(0002)\r\n0-0:96.3.10(1)\r\n0-0:17.0.0(24.000*kW)\r\n1-0:1.8.0(000113.244*kWh)\r\n1-0:1.8.1(000047.806*kWh)\r\n1-0:1.8.2(000065.438*kWh)\r\n1-0:1.8.3(000000.000*kWh)\r\n1-0:1.8.4(000000.000*kWh)\r\n ... "}
```

However, I'm not able to switch the input to any other format, like hex dump. I need this because the decoded ASCII representation seems to clamp off some non-printable characters, while the hex dump retains them in hex codes. I need them to test the CRC embedded in the output. Issuing `SSerialSend5 <string>` doesn't change it to hex dump per the [docs](https://tasmota.github.io/docs/Commands/#sserialsend). [Here](https://tasmota.github.io/docs/Commands/#serial-bridge) it also states that this should set up/change the output format, and e. g. `SerialSend5 <string>` does change it when using the HW serial bridge (I'm not using that because it doesn't support inverting RX - let's put aside the fact that on ESP32 serial is always HW, only the settings are different).

What am I overlooking?

**Comment by sfromis** [comment, marked as answer]
If you want flexibility, I'd suggest to just stop using the serial bridge, and switch the Berry, where you have full control.

Anyway, the `SerialDelimiter` command has:
> 254 = disable serial delimiter & post HEX string

**Comment by kobuki** [comment]
Thanks, `SerialDelimiter 254` did the trick. Ultimately I want Tasmota to act only as a serial to MQTT bridge after all my endeavors with the smart meter interface. I want node-red to process messages further as this gives more flexibility and readily available JS libs to process OBIS ASCII telegrams. The smart meter interface lacks a few crucial features, CRC checksumming being the most important one for me, as I have garbage in some messages and it's pretty hard to deal with downstream without a checksum.

Can we conclude that `SSerialSendX` doesn't work for setting the input format? Haven't checked any actual output yet, but I assume for that the format `X` should work.

**Reply by sfromis** [reply]
I'm not going to "conclude" anything about whether serial bridge works as it "should" or whether corrections would be called for in docs vs code. That's why I repeatedly "merely" suggest about what seems to actually get results like you ask for.

I'd still prefer to do such stuff with Berry, also in case of only wanting to emit MQTT messages with received data. A completely different approach could be https://tasmota.github.io/docs/Serial-to-TCP-Bridge/ giving you a "raw" serial tunnel with the ser2net protocol.

**Reply by kobuki** [reply]
Yeah, I get it. I'm contemplating pushing a bug report, and you seem to be well informed, hence my question.

I was also thinking about the 2 options you mentioned. Berry is nice since it's an actual programming language and has a lot of convenient stuff. But if the serial bridge works and is stable, that's another good option.

In any case, my current issue is solved, thanks.
